File: scripts/patch_caregiving_db.ts
import { createClient } from '@supabase/supabase-js';
import dotenv from 'dotenv';
import path from 'path';

// Load .env
dotenv.config({ path: path.resolve(process.cwd(), '.env') });
dotenv.config({ path: path.resolve(process.cwd(), '.env.local') });

const supabaseUrl = process.env.VITE_SUPABASE_URL || process.env.SUPABASE_URL;
const supabaseKey = process.env.SUPABASE_SERVICE_ROLE_KEY || process.env.VITE_SUPABASE_ANON_KEY;

if (!supabaseUrl || !supabaseKey) {
  console.error("[-] Supabase URL or Key not found in env.");
  process.exit(1);
}

// Use Service Role Key if available for bypass RLS and direct write permissions
const supabase = createClient(supabaseUrl, supabaseKey);

async function run() {
  console.log("[*] Fetching all rows from caregiving_insurance_plans...");
  const { data: rows, error: fetchErr } = await supabase
    .from('caregiving_insurance_plans')
    .select('*');

  if (fetchErr || !rows) {
    console.error("[-] Fetch error:", fetchErr?.message || "No data");
    return;
  }

  console.log(`[+] Fetched ${rows.length} rows.`);

  // 1. Identify dementia (치매) related rows to delete
  const dementiaIds: number[] = [];
  const pureRows: any[] = [];

  rows.forEach((row: any) => {
    const name = row.product_name || '';
    const careType = row.care_type || '';
    if (name.includes('치매') || name.includes('CDR') || careType === '치매간병') {
      dementiaIds.push(row.id);
    } else {
      pureRows.push(row);
    }
  });

  if (dementiaIds.length > 0) {
    console.log(`[*] Deleting ${dementiaIds.length} dementia-related rows...`);
    const { error: delErr } = await supabase
      .from('caregiving_insurance_plans')
      .delete()
      .in('id', dementiaIds);

    if (delErr) {
      console.error("[-] Delete error:", delErr.message);
    } else {
      console.log("[+] Successfully deleted dementia rows.");
    }
  } else {
    console.log("[*] No dementia-related rows found to delete.");
  }

  // 2. Patch pure caregiving rows
  console.log(`[*] Patching ${pureRows.length} remaining pure caregiving rows...`);
  for (const row of pureRows) {
    const malePrem = row.premium_male_40 || 0;
    const femalePrem = Math.round((malePrem * 1.2) / 10) * 10;
    
    // Determine is_increasing flag
    const name = row.product_name || '';
    let isIncreasing = false;
    if (name.includes('체증') || name.includes('RICH') || name.includes('리치') || name.includes('프리미엄') || name.includes('Rich')) {
      isIncreasing = true;
    }

    console.log(`  -> ID ${row.id} (${row.company_name} - ${name.substring(0, 20)}...): Male=${malePrem} -> Female=${femalePrem}, Increasing=${isIncreasing}`);

    const { error: updErr } = await supabase
      .from('caregiving_insurance_plans')
      .update({
        premium_female_40: femalePrem,
        is_increasing: isIncreasing
      })
      .eq('id', row.id);

    if (updErr) {
      console.error(`  [-] Update failed for ID ${row.id}:`, updErr.message);
    }
  }

  console.log("\n[*] Verification: Fetching final results...");
  const { data: finalRows, error: verifyErr } = await supabase
    .from('caregiving_insurance_plans')
    .select('id, company_name, product_name, care_type, premium_male_40, premium_female_40, is_increasing');

  if (verifyErr || !finalRows) {
    console.error("[-] Verification failed:", verifyErr?.message);
  } else {
    console.log(`[+] Total remaining pure rows: ${finalRows.length}`);
    console.log("[+] Sample of patched rows:");
    console.log(JSON.stringify(finalRows.slice(0, 5), null, 2));
  }
}

run();

File: scripts/patch_caregiving_db.py
import os
import psycopg2
from dotenv import load_dotenv

# Load environments
load_dotenv(".env.local")
load_dotenv(".env")

def run():
    db_url = os.getenv("DATABASE_URL")
    if not db_url:
        print("[-] DATABASE_URL not found in environment variables.")
        return

    try:
        conn = psycopg2.connect(db_url)
        cur = conn.cursor()
        
        # 1. Delete dementia (치매) related rows from caregiving_insurance_plans
        print("[*] Deleting dementia (치매) related rows from caregiving_insurance_plans...")
        cur.execute("""
            DELETE FROM public.caregiving_insurance_plans 
            WHERE product_name LIKE '%치매%' 
               OR care_type = '치매간병'
               OR product_name LIKE '%CDR%';
        """)
        deleted_count = cur.rowcount
        print(f"[+] Deleted {deleted_count} dementia-related rows.")

        # 2. Select remaining pure caregiving rows to patch female premiums and is_increasing flag
        cur.execute("""
            SELECT id, product_name, premium_male_40 
            FROM public.caregiving_insurance_plans;
        """)
        rows = cur.fetchall()
        print(f"[*] Found {len(rows)} remaining pure caregiving rows to patch.")

        for row in rows:
            row_id, prod_name, male_prem = row
            
            # Calculate female premium (1.2x male premium, rounded to 10 won)
            female_prem = int(round((male_prem * 1.2) / 10.0) * 10)
            
            # Determine is_increasing flag based on product name keywords
            is_increasing = False
            if any(kw in prod_name for kw in ['체증', 'RICH', '리치', '프리미엄', 'Rich', 'Premium']):
                is_increasing = True
                
            print(f"  -> Patching [ID {row_id}] {prod_name}: Male={male_prem} -> Female={female_prem}, Increasing={is_increasing}")
            
            # Update database row
            cur.execute("""
                UPDATE public.caregiving_insurance_plans 
                SET premium_female_40 = %s,
                    is_increasing = %s
                WHERE id = %s;
            """, (female_prem, is_increasing, row_id))

        # Commit transaction
        conn.commit()
        print("[+] Transaction committed successfully!")

        # 3. Print final state
        cur.execute("SELECT COUNT(*) FROM public.caregiving_insurance_plans;")
        total_cnt = cur.fetchone()[0]
        print(f"\n[+] Total rows remaining in caregiving_insurance_plans: {total_cnt}")
        
        cur.execute("SELECT id, company_name, product_name, care_type, premium_male_40, premium_female_40, is_increasing FROM public.caregiving_insurance_plans LIMIT 10;")
        colnames = [desc[0] for desc in cur.description]
        final_rows = cur.fetchall()
        print("\n[*] Final Data Samples:")
        for r in final_rows:
            print(dict(zip(colnames, r)))

        cur.close()
        conn.close()
    except Exception as e:
        print(f"[-] Database Error during patch: {e}")

if __name__ == "__main__":
    run()
